runtime-modules/project-token/src/types.rs

```diff
@@ -0,0 +1,337 @@
+    }
+
+    pub(crate) fn unclaimed_patronage<BlockNumberToBalance: Convert<BlockNumber, Balance>>(
+        &self,
+        block: BlockNumber,
+    ) -> Balance {
+        // period * rate * supply + tally
+        self.patronage_info
+            .unclaimed_patronage_percent::<BlockNumberToBalance>(block)
+            .saturating_mul(self.supply)
+            .saturating_add(self.patronage_info.unclaimed_patronage_tally_amount)
+    }
+
+    pub fn set_new_patronage_rate_at_block<BlockNumberToBalance: Convert<BlockNumber, Balance>>(
+        &mut self,
+        new_rate: Balance,
+        block: BlockNumber,
+    ) {
+        // update tally according to old rate
+        self.patronage_info.unclaimed_patronage_tally_amount =
+            self.unclaimed_patronage::<BlockNumberToBalance>(block);
+        self.patronage_info.last_unclaimed_patronage_tally_block = block;
+        self.patronage_info.rate = new_rate;
+    }
+}
+
+/// Encapsules parameters validation + TokenData construction
+impl<Balance: Zero + Copy + PartialOrd, Hash> TokenIssuanceParameters<Balance, Hash> {
+    /// Forward `self` state
+    pub fn build<BlockNumber>(self, block: BlockNumber) -> TokenData<Balance, Hash, BlockNumber> {
+        // validation
+        let patronage_info = PatronageData::<Balance, BlockNumber> {
+            last_unclaimed_patronage_tally_block: block,
+            unclaimed_patronage_tally_amount: Balance::zero(),
+            rate: self.patronage_rate,
+        };
+        TokenData::<Balance, Hash, BlockNumber> {
+            supply: self.initial_supply,
+            issuance_state: self.initial_state,
+            transfer_policy: self.transfer_policy,
+            symbol: self.symbol,
+            patronage_info,
+        }
+    }
+}
+
+impl<Hasher: Hash> MerkleProof<Hasher> {
+    pub(crate) fn verify_for_commit<T, AccountId>(
+        &self,
+        account_id: &AccountId,
+        commit: Hasher::Output,
+    ) -> DispatchResult
+    where
+        T: crate::Trait,
+        AccountId: Encode,
+    {
+        match &self.0 {
+            None => Err(crate::Error::<T>::MerkleProofNotProvided.into()),
+            Some(vec) => {
+                let init = Hasher::hash_of(account_id);
+                let proof_result = vec.iter().fold(init, |acc, (hash, side)| match side {
+                    MerkleSide::Left => Hasher::hash_of(&(hash, acc)),
+                    MerkleSide::Right => Hasher::hash_of(&(acc, hash)),
+                });
+
+                ensure!(
+                    proof_result == commit,
+                    crate::Error::<T>::MerkleProofVerificationFailure,
+                );
+
+                Ok(())
+            }
+        }
+    }
+}
+
+impl<Balance: Zero + Copy + Saturating, BlockNumber: Copy + Saturating + PartialOrd>
+    PatronageData<Balance, BlockNumber>
+{
+    pub fn unclaimed_patronage_percent<BlockNumberToBalance: Convert<BlockNumber, Balance>>(
```

Comment:
The function name is confusing. It's called`unclaimed_patronage_percent` but it returns `Balance`.
Perhaps change it to `unclaimed_patronage_since_last_rate_update`

Reply:
I removed this